File: frontend/src/api/operations.ts
/**
 * ARIA Operations Director API (Module 13 Part 7).
 *
 * The organization-scale view: aggregated organization facts, cross-farm
 * comparison, the unified operations dashboard, the merged timeline, performance
 * analytics, organization reports and task assignment. All deterministic — no
 * endpoint here reaches Gemini or Claude.
 *
 * Every aggregate carries its provenance (source farms, missing farms, method),
 * so the UI can render "not enough recorded data" honestly and never a
 * fabricated total.
 */
import apiClient from "./client";

type APISuccess<T> = { data: T; success: true };

export type Severity = "normal" | "watch" | "warning" | "critical";

export interface OpsAggregate {
  key: string;
  label: string;
  value: string | null;
  unit: string;
  available: boolean;
  source_farms: string[];
  missing_farms: string[];
  method: string;
}

export interface OpsOrgPriority {
  rank: number;
  label: string;
  why: string;
  farm_id: string;
  farm_name: string;
  severity: Severity;
  source: string;
}

export interface OpsOrganizationFacts {
  organization_name: string;
  as_of: string;
  farm_count: number;
  farm_names: string[];
  overall: Severity;
  health: OpsAggregate;
  production: OpsAggregate;
  mortality: OpsAggregate;
  feed_usage: OpsAggregate;
  water_usage: OpsAggregate;
  inventory: OpsAggregate;
  financial: OpsAggregate;
  priorities: OpsOrgPriority[];
  silent_farms: string[];
}

export interface OpsFarmRank {
  farm_id: string;
  farm_name: string;
  value: number | null;
  display: string;
  available: boolean;
}

export interface OpsRanking {
  key: string;
  label: string;
  unit: string;
  higher_is_better: boolean;
  ranked: OpsFarmRank[];
  missing: OpsFarmRank[];
  best: OpsFarmRank | null;
  needs_attention: OpsFarmRank | null;
  average: string | null;
  method: string;
}

export interface OpsComparison {
  farm_count: number;
  rankings: OpsRanking[];
}

export interface OpsWorker {
  user_id: string;
  name: string;
  role: string;
  role_label: string;
  farm_ids: string[];
  farm_names: string[];
}

export interface OpsTaskHistoryEvent {
  at: string;
  action: string;
  actor: string;
  detail: string;
}

export type TaskStatus = "open" | "overdue" | "done";

export interface OpsTask {
  task_id: string;
  farm_id: string;
  farm_name: string;
  title: string;
  status: TaskStatus;
  priority: string;
  owner_id: string | null;
  owner_name: string | null;
  due_at: string | null;
  created_at: string | null;
  completed_at: string | null;
  history: OpsTaskHistoryEvent[];
}

export interface OpsFarmSummary {
  farm_id: string;
  farm_name: string;
  overall: Severity;
  health_score: number;
  open_tasks: number;
  overdue_tasks: number;
  alert_count: number;
  silent: boolean;
}

export interface OpsAlert {
  at: string;
  farm_id: string;
  farm_name: string;
  kind: string;
  title: string;
  detail: string;
  severity: Severity;
}

export interface OpsNotification {
  at: string;
  farm_id: string;
  farm_name: string;
  title: string;
  body: string;
  severity: string;
}

export interface OpsDashboard {
  as_of: string;
  tier: string;
  sections: string[];
  organization: OpsOrganizationFacts;
  farms: OpsFarmSummary[];
  alerts: OpsAlert[];
  priorities: OpsOrgPriority[];
  tasks: OpsTask[];
  workers: OpsWorker[];
  notifications: OpsNotification[];
  operational_status: string;
}

export interface OpsTimelineEvent {
  at: string;
  farm_id: string;
  farm_name: string;
  kind: string;
  title: string;
  detail: string;
  worker: string | null;
  severity: Severity;
}

export interface OpsWorkerPerformance {
  user_id: string;
  name: string;
  assigned: number;
  completed: number;
  overdue: number;
  completion_rate: string | null;
  avg_completion_hours: string | null;
  method: string;
}

export interface OpsMetric {
  key: string;
  label: string;
  value: string | null;
  unit: string;
  available: boolean;
  method: string;
  detail: string;
}

export interface OpsAnalytics {
  workers: OpsWorkerPerformance[];
  farm_productivity: OpsMetric[];
  org_metrics: OpsMetric[];
  trends: OpsMetric[];
}

export interface OpsReportSection {
  label: string;
  value: string;
  available: boolean;
  method: string;
}

export interface OpsReport {
  period: string;
  label: string;
  as_of: string;
  organization_name: string;
  farm_count: number;
  sections: OpsReportSection[];
  risks: string[];
  priorities: string[];
  notes: string[];
}

export interface DashboardParams {
  farm_id?: string;
  worker_id?: string;
  severity?: Severity;
  since?: string;
  until?: string;
}

const base = (orgId: string) => `/organizations/${orgId}/operations`;

export async function getDashboard(orgId: string, params: DashboardParams = {}): Promise<OpsDashboard> {
  const { data } = await apiClient.get<APISuccess<OpsDashboard>>(`${base(orgId)}/dashboard`, { params });
  return data.data;
}

export async function getOrganizationFacts(orgId: string): Promise<OpsOrganizationFacts> {
  const { data } = await apiClient.get<APISuccess<OpsOrganizationFacts>>(`${base(orgId)}/organization`);
  return data.data;
}

export async function getComparison(orgId: string): Promise<OpsComparison> {
  const { data } = await apiClient.get<APISuccess<OpsComparison>>(`${base(orgId)}/farms`);
  return data.data;
}

export async function getWorkers(orgId: string): Promise<OpsWorker[]> {
  const { data } = await apiClient.get<APISuccess<OpsWorker[]>>(`${base(orgId)}/workers`);
  return data.data;
}

export async function getTasks(
  orgId: string,
  params: { farm_id?: string; worker_id?: string; include_done?: boolean } = {},
): Promise<OpsTask[]> {
  const { data } = await apiClient.get<APISuccess<OpsTask[]>>(`${base(orgId)}/tasks`, { params });
  return data.data;
}

export async function getTimeline(
  orgId: string,
  params: { days?: number; limit?: number; farm_id?: string; worker?: string; severity?: Severity } = {},
): Promise<OpsTimelineEvent[]> {
  const { data } = await apiClient.get<APISuccess<OpsTimelineEvent[]>>(`${base(orgId)}/timeline`, { params });
  return data.data;
}

export async function getAnalytics(orgId: string): Promise<OpsAnalytics> {
  const { data } = await apiClient.get<APISuccess<OpsAnalytics>>(`${base(orgId)}/analytics`);
  return data.data;
}

export async function getReport(orgId: string, period: "daily" | "weekly" | "monthly"): Promise<OpsReport> {
  const { data } = await apiClient.get<APISuccess<OpsReport>>(`${base(orgId)}/reports`, { params: { period } });
  return data.data;
}

export async function assignTask(orgId: string, taskId: string, ownerId: string): Promise<OpsTask> {
  const { data } = await apiClient.post<APISuccess<OpsTask>>(
    `${base(orgId)}/tasks/${taskId}/assign`,
    { owner_id: ownerId },
  );
  return data.data;
}

export async function completeTask(orgId: string, taskId: string): Promise<OpsTask> {
  const { data } = await apiClient.post<APISuccess<OpsTask>>(`${base(orgId)}/tasks/${taskId}/complete`, {});
  return data.data;
}

File: backend/app/schemas/operations.py
"""
Greena — Operations Director schemas (Module 13 Part 7).

Wire shapes for the organization-scale operations layer: aggregated organization
facts, cross-farm comparison, the operations dashboard, the merged timeline,
performance analytics, role-based workspace scope and organization reports.

Every aggregate mirrors the engine's honesty contract — it carries the farms that
contributed, the farms that were missing, and the calculation method — so the
client can render "not enough recorded data" and never a fabricated total.
"""

from __future__ import annotations

import uuid

from pydantic import Field

from app.schemas.base import AGRIOSSchema


# ── Provenance ────────────────────────────────────────────────────────────────


class OpsAggregate(AGRIOSSchema):
    key: str
    label: str
    value: str | None
    unit: str
    available: bool
    source_farms: list[str]
    missing_farms: list[str]
    method: str


# ── 1. Organization intelligence ──────────────────────────────────────────────


class OpsOrgPriority(AGRIOSSchema):
    rank: int
    label: str
    why: str
    farm_id: str
    farm_name: str
    severity: str
    source: str


class OpsOrganizationFacts(AGRIOSSchema):
    organization_name: str
    as_of: str
    farm_count: int
    farm_names: list[str]
    overall: str
    health: OpsAggregate
    production: OpsAggregate
    mortality: OpsAggregate
    feed_usage: OpsAggregate
    water_usage: OpsAggregate
    inventory: OpsAggregate
    financial: OpsAggregate
    priorities: list[OpsOrgPriority]
    silent_farms: list[str]


# ── 2. Cross-farm comparison ──────────────────────────────────────────────────


class OpsFarmRank(AGRIOSSchema):
    farm_id: str
    farm_name: str
    value: float | None
    display: str
    available: bool


class OpsRanking(AGRIOSSchema):
    key: str
    label: str
    unit: str
    higher_is_better: bool
    ranked: list[OpsFarmRank]
    missing: list[OpsFarmRank]
    best: OpsFarmRank | None
    needs_attention: OpsFarmRank | None
    average: str | None
    method: str


class OpsComparison(AGRIOSSchema):
    farm_count: int
    rankings: list[OpsRanking]


# ── 3/4. Workers & tasks ──────────────────────────────────────────────────────


class OpsWorker(AGRIOSSchema):
    user_id: str
    name: str
    role: str
    role_label: str
    farm_ids: list[str]
    farm_names: list[str]


class OpsTaskHistoryEvent(AGRIOSSchema):
    at: str
    action: str
    actor: str
    detail: str = ""


class OpsTask(AGRIOSSchema):
    task_id: str
    farm_id: str
    farm_name: str
    title: str
    status: str
    priority: str
    owner_id: str | None
    owner_name: str | None
    due_at: str | None
    created_at: str | None
    completed_at: str | None
    history: list[OpsTaskHistoryEvent] = Field(default_factory=list)


class OpsTaskAssign(AGRIOSSchema):
    owner_id: uuid.UUID


# ── 5. Dashboard ──────────────────────────────────────────────────────────────


class OpsFarmSummary(AGRIOSSchema):
    farm_id: str
    farm_name: str
    overall: str
    health_score: int
    open_tasks: int
    overdue_tasks: int
    alert_count: int
    silent: bool


class OpsAlert(AGRIOSSchema):
    at: str
    farm_id: str
    farm_name: str
    kind: str
    title: str
    detail: str
    severity: str


class OpsNotification(AGRIOSSchema):
    at: str
    farm_id: str
    farm_name: str
    title: str
    body: str
    severity: str


class OpsDashboard(AGRIOSSchema):
    as_of: str
    tier: str
    sections: list[str]
    organization: OpsOrganizationFacts
    farms: list[OpsFarmSummary]
    alerts: list[OpsAlert]
    priorities: list[OpsOrgPriority]
    tasks: list[OpsTask]
    workers: list[OpsWorker]
    notifications: list[OpsNotification]
    operational_status: str


# ── 6. Timeline ───────────────────────────────────────────────────────────────


class OpsTimelineEvent(AGRIOSSchema):
    at: str
    farm_id: str
    farm_name: str
    kind: str
    title: str
    detail: str
    worker: str | None
    severity: str


# ── 7. Performance analytics ──────────────────────────────────────────────────


class OpsWorkerPerformance(AGRIOSSchema):
    user_id: str
    name: str
    assigned: int
    completed: int
    overdue: int
    completion_rate: str | None
    avg_completion_hours: str | None
    method: str


class OpsMetric(AGRIOSSchema):
    key: str
    label: str
    value: str | None
    unit: str
    available: bool
    method: str
    detail: str = ""


class OpsAnalytics(AGRIOSSchema):
    workers: list[OpsWorkerPerformance]
    farm_productivity: list[OpsMetric]
    org_metrics: list[OpsMetric]
    trends: list[OpsMetric]


# ── 9. Reports ────────────────────────────────────────────────────────────────


class OpsReportSection(AGRIOSSchema):
    label: str
    value: str
    available: bool
    method: str = ""


class OpsReport(AGRIOSSchema):
    period: str
    label: str
    as_of: str
    organization_name: str
    farm_count: int
    sections: list[OpsReportSection]
    risks: list[str]
    priorities: list[str]
    notes: list[str]
